lib/semantics/expression.cc

```diff
@@ -333,7 +298,52 @@ MaybeExpr ExpressionAnalyzer::TopLevelChecks(DataRef &&dataRef) {
   return Designate(std::move(dataRef));
 }
 
+// Parse tree correction after a substring S(j:k) was misparsed as an
+// array section.  N.B. Fortran substrings have to have a range, not a
+// single index.
+static void FixMisparsedSubstring(const parser::Designator &d) {
+  auto &mutate{const_cast<parser::Designator &>(d)};
+  if (auto *dataRef{std::get_if<parser::DataRef>(&mutate.u)}) {
+    if (auto *ae{std::get_if<common::Indirection<parser::ArrayElement>>(
+            &dataRef->u)}) {
+      parser::ArrayElement &arrElement{ae->value()};
+      if (!arrElement.subscripts.empty()) {
+        auto iter{arrElement.subscripts.begin()};
+        if (auto *triplet{std::get_if<parser::SubscriptTriplet>(&iter->u)}) {
+          if (!std::get<2>(triplet->t).has_value() /* no stride */ &&
+              ++iter == arrElement.subscripts.end() /* one subscript */) {
+            if (Symbol *
+                symbol{std::visit(
+                    common::visitors{
+                        [](parser::Name &n) { return n.symbol; },
+                        [](common::Indirection<parser::StructureComponent>
+                                &sc) { return sc.value().component.symbol; },
+                        [](auto &) -> Symbol * { return nullptr; },
+                    },
+                    arrElement.base.u)}) {
+              const Symbol &ultimate{symbol->GetUltimate()};
+              if (const auto *details{
+                      ultimate.detailsIf<semantics::ObjectEntityDetails>()}) {
+                if (const semantics::DeclTypeSpec * type{details->type()}) {
+                  if (!details->IsArray() &&
```

Comment:
You could use `ultimate.GetType()` and `ultimate.IsObjectArray()` here and that would avoid the need for `details`.

Reply:
Yes, but then there would only be nine lines in a row afterwards with nothing but right braces on them.